File: C++/Basic Sensor Regulation/Basic Sensor Regulation.cpp
/*============================================================================
*               Fluigent Software Developement Kit for C++
*----------------------------------------------------------------------------
*         Copyright (c) Fluigent 2019.  All Rights Reserved.
*----------------------------------------------------------------------------
*
* Title:   Basic Sensor Regulation.cpp
* Purpose: This example shows how to set a sensor regulation and generate a
*			sinusoidal profile on the first sensor and pressure module of the chain
*
* Software: fgt_SDK_*.dll is the dynamic linked library managing all Fluigent
*			 instruments (pressure and sensors)
*			"fgt_SDK_Cpp.h" is the wrapper to fgt_SDK dll
* Hardware setup: At least one Fluigent pressure controller (MFCS, MFCS-EZ or Flow EZ)
*					and at least one Fluigent sensor (Flow Unit connected to 
*					FRP or Flow EZ)
* Version: 19.0.0.0
* Date:	06/2019
*============================================================================*/


#include <iostream>
#include <math.h>
#include "../fgt_SDK_Cpp.h"			// include wrapper to fgt_SDK dll, functions can also be accessed by loading the dll

#define PI 3.14159265

int main()
{
	// Variables declaration
	float sensorRead;				// used to read returned sensor value
	float maxSensor;				// min/max pressure controller range
	float minSensor;

	Fgt_init();						// Initialize session with all detected Fluigent instrument(s)
									// This step is optional, if not called session will be automatically created

	// Get sensor range
	Fgt_get_sensorRange(0, &minSensor, &maxSensor);

	// Read sensor value
	Fgt_get_sensorValue(0, &sensorRead);
	std::cout << "Read sensor value: " << sensorRead << std::endl;

	// Start regulation between pressure controller and sensor
	std::cout << "Start sensor regulation to " << maxSensor/10 << " (ul/min is the default unit) using first pressure controller and sensor" << std::endl;
	Fgt_set_sensorRegulation(0, 0, maxSensor / 10);

	// Wait 5 seconds for the pressure to settle
	std::cout << "Waiting 5 seconds..." << std::endl;
	Sleep(5000);

	// Read sensor value
	Fgt_get_sensorValue(0, &sensorRead);
	std::cout << "Read sensor value: " << sensorRead << std::endl;

	// Create a sinusoidal profile from 0 through maximal and minimal sensor range
	std::cout << "Set sinusoidal orders from 0 through maximal and minimal sensor range" << std::endl;
	for (int loop = 0; loop < 360; loop+=10)
	{
		// Set regulation
		std::cout << "Set sensor regulation: " << sin(loop * 2 * PI / 360) * maxSensor << std::endl;
		Fgt_set_sensorRegulation(0, 0, sin(loop * 2 * PI / 360) * maxSensor);

		// Wait 1 second
		Sleep(1000);

		// Read sensor value
		Fgt_get_sensorValue(0, &sensorRead);
		std::cout << "Read sensor value: " << sensorRead << std::endl;
	}

	Fgt_set_pressure(0, 0);	// Set pressure to 0 before closing, it also stops running regulation

	Fgt_close();								// Close session

	system("PAUSE");
}

File: Python/examples/BasicSensorRegulation.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Basic Sensor Regulation example

This example shows how to set a sensor regulation and generate a
sinusoidal profile on the first sensor and pressure module of the chain
			
Requires at least one Fluigent pressure controller (MFCS, MFCS-EZ or
Flow EZ) and at least one Fluigent sensor (Flow Unit connected to FRP or
Flow EZ)


Copyright (c) Fluigent 2019.  All Rights Reserved.
"""

# Print function for Python 2 compatibility
from __future__ import print_function 
import time
from math import sin, pi

from Fluigent.SDK import fgt_init, fgt_close
from Fluigent.SDK import fgt_get_sensorValue, fgt_get_sensorRange
from Fluigent.SDK import fgt_set_sensorRegulation, fgt_set_pressure

## Initialize the session
# This step is optional, if not called session will be automatically
# created
fgt_init()

## Get sensor range   
minSensor, maxSensor = fgt_get_sensorRange(0)

## Read sensor value
sensorMeasurement = fgt_get_sensorValue(0)
print('Current sensor value: {}'.format(sensorMeasurement))

# Set regulation setpoint to 10% of the maximum sensor value
# Use the first sensor and the first pressure channel detected
fgt_set_sensorRegulation(0, 0, maxSensor/10)

# Wait 5 seconds for the pressure to settle
print('Waiting 5 seconds...')
time.sleep(5)

# Read sensor value
sensorMeasurement = fgt_get_sensorValue(0)
print('Current sensor value: {:.2f}'.format(sensorMeasurement))

## Create a sinusoidal profile 
# The profile oscillates between the minimum and the maximum sensor value

# Create the sine wave setpoints
amplitude = (maxSensor-minSensor)
offset = minSensor
sinePoints = [(sin(angle*pi/180)+1)/2 for angle in range(0, 360, 10)]
setpoints = [value*amplitude + offset for value in sinePoints]

for sensorSetpoint in setpoints:
    # Set new setpoint
    print('Set sensor regulation to {:.2f}'.format(sensorSetpoint))
    fgt_set_sensorRegulation(0, 0, sensorSetpoint)
    # Wait 1 second
    time.sleep(1)
    # Read sensor value
    sensorMeasurement = fgt_get_sensorValue(0)
    print('Current sensor value: {:.2f}'.format(sensorMeasurement))

## Close the session
# Set pressure to 0 before closing. This also stops the regulation
fgt_set_pressure(0, 0)
fgt_close()
